persistence/sql/publish_to_local_queue.go

```diff
@@ -0,0 +1,73 @@
+// Copyright 2023 XDBLab organization
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sql
+
+import (
+	"context"
+	"github.com/xdblab/xdb/common/log/tag"
+	"github.com/xdblab/xdb/extensions"
+	"github.com/xdblab/xdb/persistence"
+)
+
+func (p sqlProcessStoreImpl) PublishToLocalQueue(ctx context.Context, request persistence.PublishToLocalQueueRequest) (
+	*persistence.PublishToLocalQueueResponse, error) {
+	tx, err := p.session.StartTransaction(ctx, defaultTxOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := p.doPublishToLocalQueueTx(ctx, tx, request)
+
+	if err != nil {
+		err2 := tx.Rollback()
+		if err2 != nil {
+			p.logger.Error("error on rollback transaction", tag.Error(err2))
+		}
+	} else if resp.HasNewImmediateTask { // if there is no NewImmediateTask, it means there is nothing to commit in the transaction
```

Comment:
@longquanzheng It seems if there is no actual sql operation, commit will fail.

Reply:
No operation do you mean because the sql doesn’t have any effect because of duplicates?

Reply:
Yes, if commit, the error message is `failed to commit on failure transaction`.

Reply:
Modified!